common/config.go

```diff
@@ -144,27 +182,38 @@ func DownloadableURL(original string) (string, error) {
 
 func FileExistsLocally(original string) bool {
 	// original should be something like file://C:/my/path.iso
+	u, _ := url.Parse(original)
+
+	// First create a dummy downloader so we can figure out which
+	// protocol to use.
+	cli := NewDownloadClient(&DownloadConfig{})
+	d, ok := cli.config.DownloaderMap[u.Scheme]
+	if !ok {
+		return false
+	}
 
-	fileURL, _ := url.Parse(original)
-	fileExists := false
-
-	if fileURL.Scheme == "file" {
-		// on windows, correct URI is file:///c:/blah/blah.iso.
-		// url.Parse will pull out the scheme "file://" and leave the path as
-		// "/c:/blah/blah/iso".  Here we remove this forward slash on absolute
-		// Windows file URLs before processing
-		// see https://blogs.msdn.microsoft.com/ie/2006/12/06/file-uris-in-windows/
-		// for more info about valid windows URIs
-		filePath := fileURL.Path
-		if runtime.GOOS == "windows" && len(filePath) > 0 && filePath[0] == '/' {
-			filePath = filePath[1:]
-		}
-		_, err := os.Stat(filePath)
-		if err != nil {
-			return fileExists
-		} else {
-			fileExists = true
-		}
+	// Check to see that it's got a Local way of doing things.
+	local, ok := d.(LocalDownloader)
```

Comment:
It looks to me like you've changed part of the functionality of the original implementation; if the URL scheme isn't "file" (e.g. this is to be downloaded from the internet) then we return `true`; here it looks like if the URL scheme isn't "file" it'll return `false`. Am I reading this wrong?

Reply:
So the logic for this, is pretty much the following:

1. Convert path to a net/url.URL
2. Create a Downloader (so that we can find which schemes support a local path). If we can't, then `fail`.
3. Grab the scheme (or protocol) from the url and check if there's a Downloader for it. If one doesn't exist, then `return false`.
4. Is this a locally accessible protocol (by checking the downloader implements the `LocalDownloader` interface)? If it's not, then `return false`.
5. Figure out the real path to the file using LocalDownloader's `.toPath`. If we error here, then `return false`.
6. Call `os.Stat()` on it. If we error, then file is non-accessible..so, `return false`.
7. We succeeded, so the file pointed to by the url must exist.

This requires protocols that support local paths to implement a `.toPath()` method that returns the local path for a uri. Right now, only the "file" and "smb" downloaders implement the `LocalDownloader` interface and so these are the only protocols that get past step 4. This way if someone wants to implement the afp:// protocol for osx for some reason, they can choose whether to support a local path transformation or not.

Anything that doesn't implement that interface (e.g. things to be downloaded from the internet) will cause step 4 to return false.

Reply:
> Anything that doesn't implement that interface (e.g. things to be downloaded from the internet) will cause step 4 to return false.

Right, that's the logic change I was asking about. In the original logic, something that should be downloaded from the internet wouldn't return false, it'd return true.  See the function docstring for the intended behavior -- essentially, it should only return false if both a) it is supposed to be found locally (e.g. it's using the file:// protocol) AND b) the file is not actually available locally.

Reply:
Ahh. Ok. Gots it. I'll invert that case then.

Reply:
Interesting when I developed this, how the tests in common/config_test.go actually passed. I discovered why and fixed it. Let's see what Travis-CI says...